bin/benchmark_performance.rb

```diff
@@ -0,0 +1,425 @@
+    return h
+  end
+
+  rd, wr = IO.pipe
+  p1 = fork do
+    h = {}
+    block.call(h)
+    wr.write(JSON.dump(h))
+    wr.close
+  end
+  Process.wait(p1)
+  wr.close
+  h = JSON.parse(rd.read, symbolize_names: true)
+  rd.close
+  return h
+end
+
+def benchmark_require(gem, data)
+  data.merge!(fork_run do |out|
+    t1 = monotonic_milliseconds
+    require gem
+    out[:require_time_ms] = (monotonic_milliseconds - t1)
+  end)
+
+  data.merge!(fork_run do |out|
+    unless defined?(JRUBY_VERSION)
+      r = ::MemoryProfiler.report { require gem }
+      out[:require_mem_retained_kb] = r.total_retained_memsize / 1024.0
+      out[:require_mem_allocated_kb] = r.total_allocated_memsize / 1024.0
+    end
+  end)
+end
+
+def benchmark_client(gem, module_name, data)
+  data.merge!(fork_run do |out|
+    require gem
+    client_klass = Aws.const_get(module_name).const_get(:Client)
+    unless defined?(JRUBY_VERSION)
+      r = ::MemoryProfiler.report { client_klass.new(stub_responses: true) }
+      out[:client_mem_retained_kb] = r.total_retained_memsize / 1024.0
+      out[:client_mem_allocated_kb] = r.total_allocated_memsize / 1024.0
+    end
+  end)
+end
+
+# This runs in the main process and requires service gems.
+# It MUST be done after ALL testing of gem loads/client creates
+def run_benchmarks(gem, module_name, benchmarks, data)
+  require gem
+  client_klass = Aws.const_get(module_name).const_get(:Client)
+
+  data[:client_init_ms] = benchmark(300) do
+    client_klass.new(stub_responses: true)
+  end
+  values = data[:client_init_ms]
+  puts "\t\t#{gem} client init avg: #{'%.2f' % (values.sum(0.0) / values.size)} ms"
+
+  benchmarks.each do |test_name, test_def|
+    client = client_klass.new(stub_responses: true)
+    req = test_def[:setup].call(client)
+    n = test_def[:n] || 300
+    # cloudwatch has a values limit of 150
+    values = benchmark(n) do
+      test_def[:test].call(client, req)
+    end
+    data["#{test_name}_ms"] = values
+    puts "\t\t#{test_name} avg: #{ '%.2f' % (values.sum(0.0) / values.size)} ms"
+  end
+end
+
+def host_os
+  case RbConfig::CONFIG['host_os']
+  when /mac|darwin/
+    'macos'
+  when /linux|cygwin/
+    'linux'
+  when /mingw|mswin/
+    'windows'
+  else
+    'other'
+  end
+end
+
+# polyfill random_number for older versions of Ruby.
+unless Random.new.respond_to?(:random_number)
+  class Random
+    def random_number(n)
+      (rand*n).floor
+    end
+  end
+end
+
+def random_value(i=0, seed=0)
+  r = Random.new(i+seed) # use the index as the seed for predictable results
+  case r.random_number(5)
+  when 0 then "Some String value #{i}"
+  when 1 then r.rand # a float value
+  when 2 then r.random_number(100000) # a large integer
+  when 3 then (0..r.random_number(100)).to_a # an array
+  when 4 then {a: 1, b: 2, c: 3} # a hash
+  else
+    "generic string"
+  end
+end
+
+def test_hash(n_keys=5, seed=0)
+  n_keys.times.map { |i| ["key#{i}", random_value(i, seed)]}.to_h
+end
+
+benchmark_gems = {
```

Comment:
It'd be nice to define this somewhere else, like in a yaml

Reply:
Yeah - I agree - I'd like a nice way to define this and split it up.  I don't think yaml would work well since its a lot of actual code (ie, blocks - and those blocks need functions defined, ect). Any other thoughts on how to break this down and organize it better?